### PR TITLE
lua: Expose stream info downstreamLocalAddress and downstreamDirectRemoteAddress for Lua filter

### PR DESCRIPTION
**Pull request by nic-chen**
Commit Message:  
Client IP should be very commonly used and helpful for users.
Expose `downstreamLocalAddress` and `downstreamDirectRemoteAddress` help us get client IP in  Lua filters.

Risk Level: Low
Testing: Modify unit test
Docs Changes: updated Lua filter doc
Release Notes: N/A
Fixes https://github.com/envoyproxy/envoy/issues/13457

Signed-off-by: nic-chen johz@163.com

### CODE REVIEW

File: test/extensions/filters/http/lua/wrappers_test.cc [hunk 1]
```diff
@@ -269,6 +271,50 @@ TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentProtocol) {
...
+}
+
+// Verify downstream direct remote addresses are available from stream info wrapper.
+TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentDownstreamAddresses) {
+  const std::string SCRIPT{R"EOF(
+      function callMe(object)
+        testPrint(string.format("%s", object:downstreamDirectRemoteAddress()))
+      end
+    )EOF"};
+
```
Comment (dio) [inline test/extensions/filters/http/lua/wrappers_test.cc:303]: Could you merge this to the above test?

```lua
function callMe(object)
  testPrint(object:downstreamLocalAddress()))
  testPrint(object: downstreamDirectRemoteAddress())
end
```

And you can mock different address for each, for example:

```
auto downstream_local = Network::Address::InstanceConstSharedPtr{
       new Network::Address::Ipv4Instance("127.0.0.1", 8000)};
auto downstream_direct_remote = Network::Address::InstanceConstSharedPtr{
       new Network::Address::Ipv4Instance("8.8.8.8", 3000)};

```

**Review by dio — CHANGES_REQUESTED** [review]
Thanks, looks good. 

I think you need to add an entry to `docs/root/version_history/current.rst` telling that we are adding these two new APIs as "New Features" (there is a couple of examples of Lua changes entries in, e.g.` docs/root/version_history/v1.8.0.rst`)

File: test/extensions/filters/http/lua/wrappers_test.cc (hunk 1 above)
Comment (nic-chen) [reply test/extensions/filters/http/lua/wrappers_test.cc:303]: sure, thanks.

Comment (nic-chen) [reply test/extensions/filters/http/lua/wrappers_test.cc]: I tried to merge into one test case, but ran fail, even though mocked a same IP. They all run ok when separated.

I wonder if they can't be mocked together or something wrong? Please take a look. Thanks !

here is the err log:

```
[       OK ] LuaStreamInfoWrapperTest.SetGetComplexDynamicMetadata (1 ms)
                                                                                                                                      117,2         83%
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: called once - saturated and active
test/extensions/filters/http/lua/wrappers_test.cc:253: tried expectation #11: EXPECT_CALL(printer_, testPrint(fmt::format("'{}'", Http::Utility::getProtocolString(protocol.value()))))...
  Expected arg #0: is equal to "'HTTP/1.1'"
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: called once - saturated and active
test/extensions/filters/http/lua/wrappers_test.cc:253: tried expectation #12: EXPECT_CALL(printer_, testPrint(fmt::format("'{}'", Http::Utility::getProtocolString(protocol.value()))))...
  Expected arg #0: is equal to "'HTTP/2'"
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: called once - saturated and active
test/extensions/filters/http/lua/wrappers_test.cc:295: tried expectation #13: EXPECT_CALL(printer_, testPrint(address->asString()))...
  Expected arg #0: is equal to "127.0.0.1:8000"
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: called once - saturated and active
test/extensions/filters/http/lua/wrappers_test.cc:296: tried expectation #14: EXPECT_CALL(printer_, testPrint(downstream_direct_remote->asString()))...
  Expected arg #0: is equal to "127.0.0.1:8000"
           Actual: "127.0.0.1:0"
         Expected: to be called once
           Actual: never called - unsatisfied and active
Stack trace:
  0x2bbbfb9: testing::internal::GoogleTestFailureReporter::ReportFailure()
  0x12450a8: testing::internal::Expect()
  0x2bbe685: testing::internal::UntypedFunctionMockerBase::UntypedInvokeWith()::$_1::operator()()
  0x2bbe2b7: testing::internal::UntypedFunctionMockerBase::UntypedInvokeWith()
  0x1200c74: testing::internal::FunctionMocker<>::Invoke()
  0x1200c25: Envoy::Extensions::Filters::Common::Lua::Printer::testPrint()
  0x122ce43: Envoy::Extensions::Filters::Common::Lua::LuaWrappersTestBase<>::luaTestPrint()
  0x12c7a5a: lj_BC_FUNCC
```

Comment (dio) [reply test/extensions/filters/http/lua/wrappers_test.cc]: Sorry, I think using `ON_CALL` would help:

```diff
diff --git a/test/extensions/filters/http/lua/wrappers_test.cc b/test/extensions/filters/http/lua/wrappers_test.cc
index 81c38f6de..9cdacee4b 100644
--- a/test/extensions/filters/http/lua/wrappers_test.cc
+++ b/test/extensions/filters/http/lua/wrappers_test.cc
@@ -276,8 +276,8 @@ TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentProtocol) {
 TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentDownstreamAddresses) {
   const std::string SCRIPT{R"EOF(
       function callMe(object)
-        testPrint(string.format("%s", object:downstreamLocalAddress()))
-        testPrint(string.format("%s", object:downstreamDirectRemoteAddress()))
+        testPrint(string.format(object:downstreamLocalAddress()))
+        testPrint(string.format(object:downstreamDirectRemoteAddress()))
       end
     )EOF"};
 
@@ -289,9 +289,9 @@ TEST_F(LuaStreamInfoWrapperTest, ReturnCurrentDownstreamAddresses) {
       new Network::Address::Ipv4Instance("127.0.0.1", 8000)};
   auto downstream_direct_remote =
       Network::Address::InstanceConstSharedPtr{new Network::Address::Ipv4Instance("8.8.8.8", 3000)};
-  EXPECT_CALL(stream_info, downstreamLocalAddress()).WillRepeatedly(ReturnRef(address));
-  EXPECT_CALL(stream_info, downstreamDirectRemoteAddress())
-      .WillRepeatedly(ReturnRef(downstream_direct_remote));
+  ON_CALL(stream_info, downstreamLocalAddress()).WillByDefault(ReturnRef(address));
+  ON_CALL(stream_info, downstreamDirectRemoteAddress())
+      .WillByDefault(ReturnRef(downstream_direct_remote));
   Filters::Common::Lua::LuaDeathRef<StreamInfoWrapper> wrapper(
       StreamInfoWrapper::create(coroutine_->luaState(), stream_info), true);
   EXPECT_CALL(printer_, testPrint(address->asString()));
```

**Comment by dio** [comment]
Add @nic-chen, sorry, after applying my suggestion here: https://github.com/envoyproxy/envoy/pull/13536#discussion_r506044523 could you merge main?  🙏🏽

File: test/extensions/filters/http/lua/wrappers_test.cc (hunk 1 above)
Comment (nic-chen) [reply test/extensions/filters/http/lua/wrappers_test.cc]: @dio 
it works!  thanks for your help!  It saves me from going too far on the wrong way.

**Comment by nic-chen** [comment]
> @nic-chen, sorry, after applying my suggestion here: [#13536 (comment)](https://github.com/envoyproxy/envoy/pull/13536#discussion_r506044523) could you merge main? 🙏🏽

done.

File: docs/root/version_history/current.rst [hunk 2]
```diff
@@ -30,6 +30,8 @@ New Features
...
 * hds: added support for delta updates in the :ref:`HealthCheckSpecifier <envoy_v3_api_msg_service.health.v3.HealthCheckSpecifier>`, making only the Endpoints and Health Checkers that changed be reconstructed on receiving a new message, rather than the entire HDS.
 * health_check: added option to use :ref:`no_traffic_healthy_interval <envoy_v3_api_field_config.core.v3.HealthCheck.no_traffic_healthy_interval>` which allows a different no traffic interval when the host is healthy.
+* lua: added :ref:`streamInfo():downstreamLocalAddress() <config_http_filters_lua_stream_info_wrapper>` API.
+* lua: added :ref:`streamInfo():downstreamDirectRemoteAddress() <config_http_filters_lua_stream_info_wrapper>` API.
```
Comment (dio) [inline docs/root/version_history/current.rst:34]: I think this can go in one line

**Comment by dio** [comment]
@nic-chen I think you can do merge master instead of rebasing your branch with the master (so you don't have to do force-pushing).

**Comment by nic-chen** [comment]
@dio 
OK. thanks for tips.
will pay attention next time.
do I need to do something now ?

File: docs/root/version_history/current.rst (hunk 2 above)
Comment (nic-chen) [reply docs/root/version_history/current.rst]: ok.

File: docs/root/version_history/current.rst [hunk 3]
```diff
@@ -30,6 +30,7 @@ New Features
...
 * hds: added support for delta updates in the :ref:`HealthCheckSpecifier <envoy_v3_api_msg_service.health.v3.HealthCheckSpecifier>`, making only the Endpoints and Health Checkers that changed be reconstructed on receiving a new message, rather than the entire HDS.
 * health_check: added option to use :ref:`no_traffic_healthy_interval <envoy_v3_api_field_config.core.v3.HealthCheck.no_traffic_healthy_interval>` which allows a different no traffic interval when the host is healthy.
+* lua: added `downstreamDirectRemoteAddress()` and `downstreamLocalAddress()` APIs :ref:`streamInfo() <config_http_filters_lua_stream_info_wrapper>`.
 * listener: added an optional :ref:`default filter chain <envoy_v3_api_field_config.listener.v3.Listener.default_filter_chain>`. If this field is supplied, and none of the :ref:`filter_chains <envoy_v3_api_field_config.listener.v3.Listener.filter_chains>` matches, this default filter chain is used to serve the connection.
```
Comment (dio) [inline docs/root/version_history/current.rst:34]: Should swap these two lines (sort it alphabetically). That's why the CI is complaining: https://dev.azure.com/cncf/4684fb3d-0389-4e0b-8251-221942316e06/_apis/build/builds/54458/logs/11

If you run:

```
./tools/code_format/check_format.py fix ./docs/root/version_history/current.rst
```

you will get:

```
ERROR: From ././docs/root/version_history/current.rst
ERROR: ././docs/root/version_history/current.rst:34: Version history not in alphabetical order (lua vs listener): please check placement of line
 * listener: added an optional :ref:`default filter chain <envoy_v3_api_field_config.listener.v3.Listener.default_filter_chain>`. If this field is supplied, and none of the :ref:`filter_chains <envoy_v3_api_field_config.listener.v3.Listener.filter_chains>` matches, this default filter chain is used to serve the connection..
```

Comment (nic-chen) [reply docs/root/version_history/current.rst]: OK, fixed.

**Review by dio — DISMISSED** [review]
Thanks!

File: docs/root/version_history/current.rst [hunk 4]
```diff
@@ -32,6 +32,7 @@ New Features
 * hds: added support for delta updates in the :ref:`HealthCheckSpecifier <envoy_v3_api_msg_service.health.v3.HealthCheckSpecifier>`, making only the Endpoints and Health Checkers that changed be reconstructed on receiving a new message, rather than the entire HDS.
 * health_check: added option to use :ref:`no_traffic_healthy_interval <envoy_v3_api_field_config.core.v3.HealthCheck.no_traffic_healthy_interval>` which allows a different no traffic interval when the host is healthy.
 * listener: added an optional :ref:`default filter chain <envoy_v3_api_field_config.listener.v3.Listener.default_filter_chain>`. If this field is supplied, and none of the :ref:`filter_chains <envoy_v3_api_field_config.listener.v3.Listener.filter_chains>` matches, this default filter chain is used to serve the connection.
+* lua: added `downstreamDirectRemoteAddress()` and `downstreamLocalAddress()` APIs :ref:`streamInfo() <config_http_filters_lua_stream_info_wrapper>`.
```
Comment (dio) [inline docs/root/version_history/current.rst:35]: Oops, sorry, missing "to"?

lua: added `downstreamDirectRemoteAddress()` and `downstreamLocalAddress()` APIs to :ref:`stream info object <config_http_filters_lua_stream_info_wrapper>`.

Comment (nic-chen) [reply docs/root/version_history/current.rst]: fixed. thanks!

**Review by dio — APPROVED** [review]
Thanks!

**Review by mattklein123 — APPROVED** [review]

**Pull request merged**